### PR TITLE
further changes to *com notation (summarized below)

### PR DESCRIPTION
**Pull request by jonsterling**
+ use "system"-style notation: `hcom{r~>r'}(A; cap) [tubes]`
+ don't bother to support old "abstract" style notation: the fancy notation is always easier to read and write
+ in the *concrete* syntax, write `fcom` instead of `fhcom`; we will change the abstract syntax later (I could do it now, just wanted to be sure not to interfere with other changes that are being worked on)

### CODE REVIEW

File: test/success/connection.prl [hunk 1]
```diff
@@ -7,13 +7,20 @@ Thm Connection(#A) : [
...
+             k=1 -> {_}.,a];
+         i=1 -> {k}. hcom{1~>j}(#A; ,p @ k) [k=0 -> {_}.,a; k=1 -> {l}.,p @ l];
+         j=0 -> {k}.
+          hcom{1~>i}(#A; hcom{1~>0}(#A; ,p @ k) [k=0 -> {_}.,a; k=1 -> {l}.,p @ l])
+            [k=0 -> {_}.,a;
+             k=1 -> {_}.,a];
+         j=1 -> {k}. hcom{1~>i}(#A; ,p @ k) [k=0 -> {_}.,a; k=1 -> {l}.,p @ l]]
   };
-  auto;
+  auto
```
Comment (jonsterling) [inline test/success/connection.prl]: BTW, I tested this file with both SML/NJ and MLton; in SML/NJ it takes .76s to refine, whereas with MLton it takes only .46s to refine.

**Comment by favonia** [comment]
@jonsterling Hmm some of my comments seemed to disappear. Let me retype them. I was thinking whether we should have
```
comOperatorHead LBRACKET dir RBRACKET LPAREN bindings SEMI tubes RPAREN
```
as a unified rule. `comOperatorHead` returns `HCOM` or `FCOM`. They are currently separate because I wanted to reduce shift/reduce conflicts, but now with `[` and `]` it is unnecessary.

**Comment by favonia** [comment]
@jonsterling Alright. I found my comment and your response: https://github.com/RedPRL/sml-redprl/commit/03daecb7aad0bd9d98e87d578215a3bedeafaf84#commitcomment-22685220

**Comment by jonsterling** [comment]
FYI, if you comment on a commit, it can disappear from the record if that commit is rebased; but if you comment on the diff of the pull request itself, the comment is persistent, I think.

File: test/success/connection.prl (hunk 1 above)
Comment (favonia) [inline test/success/connection.prl]: I did some study but decided not to spend too much time now. You might be interested in this:
```
0.31 seconds of CPU time (0.13 seconds GC)
            function              cur 
-------------------------------- -----
<gc>                             29.5%
Telescope.Internal.ConsView.out  22.7%
Susp.delay                        6.8%
Abt.check                         4.5%
Abt.inferb                        4.5%
Abt.infer                         2.3%
Abt.symctx                        2.3%
Abt.liberateSymbolAnn             2.3%
Abt.liberateVariable.fn           2.3%
Abt.foldStar.fn.fn                2.3%
Abt.ShiftFunCat.cmp               2.3%
Abt.assertSortEq                  2.3%
Susp.force                        2.3%
DictFun.union                     2.3%
SplayRDict.foldl                  2.3%
AbtParameterTerm.Monad.bind.fn    2.3%
SplayTree.findAndSplay            2.3%
RedPrlCategoricalJudgment.map.fn  2.3%
Sequence.append                   2.3%
```

Comment (jonsterling) [inline test/success/connection.prl]: Very interesting! I bet we can improve this by making various part of the implementation less crappy... Thanks for the info.

Comment (jonsterling) [inline test/success/connection.prl]: But yeah, it's probably not a huge priority at the moment, until we hit some performance walls.

**Pull request merged**